From 1929112f055681eca1681d8dd1ba30558570e750 Mon Sep 17 00:00:00 2001
From: Eran Cohen <eran@stratoscale.com>
Date: Sun, 25 Oct 2020 15:07:57 +0200
Subject: [PATCH 1/2] Added patch.service patch the cluster with overrides for
 single node during the installation Currently the service needs to run while
 cluster-bootstrap is running (control plane is up).

THIS IS A TEMP HACK ALLOWING SINGLE NODE INSTALLATION
This is just for the POC and will be replaced by
cluster-etcd-operator and cluster-authentication-operator changes, we have kept it in the bootstrap phase for now.
See https://issues.redhat.com/browse/ETCD-155 https://issues.redhat.com/browse/MSTR-1125
---
 .../bootstrap/files/usr/local/bin/patch.sh    | 33 +++++++++++++++++++
 .../bootstrap/systemd/units/patch.service     | 15 +++++++++
 pkg/asset/ignition/bootstrap/bootstrap.go     |  1 +
 3 files changed, 49 insertions(+)
 create mode 100755 data/data/bootstrap/files/usr/local/bin/patch.sh
 create mode 100644 data/data/bootstrap/systemd/units/patch.service

diff --git a/data/data/bootstrap/files/usr/local/bin/patch.sh b/data/data/bootstrap/files/usr/local/bin/patch.sh
new file mode 100755
index 000000000..224cefa38
--- /dev/null
+++ b/data/data/bootstrap/files/usr/local/bin/patch.sh
@@ -0,0 +1,33 @@
+#!/bin/bash -x
+
+function patchit {
+    # allow etcd-operator to start the etcd cluster without minimum of 3 master nodes
+    oc --kubeconfig ./auth/kubeconfig patch etcd cluster -p='{"spec": {"unsupportedConfigOverrides": {"useUnsupportedUnsafeNonHANonProductionUnstableEtcd": true}}}' --type=merge || return 1
+
+    # allow cluster-authentication-operator to deploy OAuthServer without minimum of 3 master nodes
+    oc --kubeconfig ./auth/kubeconfig patch authentications.operator.openshift.io cluster -p='{"spec": {"managementState": "Managed", "unsupportedConfigOverrides": {"useUnsupportedUnsafeNonHANonProductionUnstableOAuthServer": true}}}' --type=merge || return 1
+
+    # Mark etcd-quorum-guard as unmanaged
+    oc --kubeconfig ./auth/kubeconfig patch clusterversion/version --type='merge' -p "$(cat <<- EOF
+ spec:
+    overrides:
+      - group: apps/v1
+        kind: Deployment
+        name: etcd-quorum-guard
+        namespace: openshift-etcd
+        unmanaged: true
+EOF
+)" || return 1
+
+    # scale down etcd-quorum-guard
+    oc --kubeconfig ./auth/kubeconfig scale --replicas=0 deployment/etcd-quorum-guard -n openshift-etcd || return 1
+
+    return 0
+}
+
+while ! patchit; do
+    echo "Waiting to try again..."
+    sleep 10
+done
+touch patch.done
+
diff --git a/data/data/bootstrap/systemd/units/patch.service b/data/data/bootstrap/systemd/units/patch.service
new file mode 100644
index 000000000..de0aa2477
--- /dev/null
+++ b/data/data/bootstrap/systemd/units/patch.service
@@ -0,0 +1,15 @@
+[Unit]
+Description=Patch single node OpenShift cluster
+Wants=bootkube.service
+ConditionPathExists=!/opt/openshift/patch.done
+
+[Service]
+WorkingDirectory=/opt/openshift
+ExecStart=/usr/local/bin/patch.sh
+
+Restart=on-failure
+RestartSec=5s
+
+
+[Install]
+WantedBy=multi-user.target
\ No newline at end of file
diff --git a/pkg/asset/ignition/bootstrap/bootstrap.go b/pkg/asset/ignition/bootstrap/bootstrap.go
index cd342d07a..ce39dd2ee 100644
--- a/pkg/asset/ignition/bootstrap/bootstrap.go
+++ b/pkg/asset/ignition/bootstrap/bootstrap.go
@@ -372,6 +372,7 @@ func (a *Bootstrap) addSystemdUnits(uri string, templateData *bootstrapTemplateD
 		"chown-gatewayd-key.service":      {},
 		"systemd-journal-gatewayd.socket": {},
 		"approve-csr.service":             {},
+		"patch.service":             {},
 		// baremetal & openstack platform services
 		"keepalived.service":        {},
 		"coredns.service":           {},
-- 
2.24.1

